python/feathub/processors/local/sliding_window_utils.py

```diff
@@ -0,0 +1,353 @@
+        self.expr = expr
+        self.agg_func = agg_func
+        self.window_size = window_size
+        self.filter_expr = filter_expr
+        self.limit = limit
+
+    @staticmethod
+    def from_feature(feature: Feature) -> "AggregationFieldDescriptor":
+        transform = feature.transform
+        if not (isinstance(transform, SlidingWindowTransform)):
+            raise FeathubException(
+                f"Cannot convert {feature} to AggregationFieldDescriptor."
+            )
+        return AggregationFieldDescriptor(
+            feature.name,
+            to_numpy_dtype(feature.dtype),
+            transform.expr,
+            transform.agg_func,
+            transform.window_size,
+            transform.filter_expr,
+            transform.limit,
+        )
+
+
+class SlidingWindowDescriptor:
+    """
+    Descriptor of a sliding window.
+    """
+
+    def __init__(
+        self,
+        step_size: timedelta,
+        group_by_keys: Sequence[str],
+    ) -> None:
+        self.step_size = step_size
+        self.group_by_keys = group_by_keys
+
+    @staticmethod
+    def from_sliding_window_transform(
+        sliding_window_agg: SlidingWindowTransform,
+    ) -> "SlidingWindowDescriptor":
+        return SlidingWindowDescriptor(
+            sliding_window_agg.step_size,
+            sliding_window_agg.group_by_keys,
+        )
+
+    def __eq__(self, other: object) -> bool:
+        return (
+            isinstance(other, self.__class__)
+            and self.step_size == other.step_size
+            and self.group_by_keys == other.group_by_keys
+        )
+
+    def __hash__(self) -> int:
+        return hash(
+            (
+                self.step_size,
+                tuple(self.group_by_keys),
+            )
+        )
+
+
+def evaluate_sliding_window(
+    input_df: pd.DataFrame,
+    feature_view: SlidingFeatureView,
+    window_descriptor: SlidingWindowDescriptor,
+    agg_descriptors: List[AggregationFieldDescriptor],
+    tz: tzinfo,
+    parser: ExprParser,
+    ast_evaluator: LocalAstEvaluator,
+) -> pd.DataFrame:
+    """
+    Evaluate the sliding window on the input DataFrame.
+    """
+    df_copy = input_df.copy()
+    append_unix_time_column(
+        df_copy,
+        feature_view.get_resolved_source().timestamp_field,
+        feature_view.get_resolved_source().timestamp_format,
+        tz,
+    )
+    df_copy[EVENT_TIME_ATTRIBUTE_NAME] = (
+        df_copy[EVENT_TIME_ATTRIBUTE_NAME] * 1000
+    ).astype(np.int)
+
+    if len(window_descriptor.group_by_keys) > 0:
+        agg_df = df_copy.groupby(by=window_descriptor.group_by_keys).apply(
+            _get_sliding_window_func(
+                sliding_window_descriptor=window_descriptor,
+                agg_field_descriptors=agg_descriptors,
+                tz=tz,
+                parser=parser,
+                ast_evaluator=ast_evaluator,
+            )
+        )
+    else:
+        agg_df = df_copy.sort_values(by=[EVENT_TIME_ATTRIBUTE_NAME])
```

Comment:
Do we also need to sort df by EVENT_TIME_ATTRIBUTE_NAME when `len(window_descriptor.group_by_keys) > 0`?